docs/adr/0004-only-threadsafe-interfaces.md

```diff
@@ -0,0 +1,156 @@
+
+* Opens the door to enhancements that would be impossible for non-`Send+Sync`
+  interfaces, and simpler to implement for `Send+Sync` interfaces if support
+  for non-`Send+Sync` interfaces did not exist.
+
+* Simpler implementation and documentation.
+
+### Negative Consequences
+
+* All consumers (both inside Mozilla and external) will need to change their
+  interfaces to be `Send+Sync`. As an example of what this entails,
+  see [this commit](https://github.com/mozilla/uniffi-rs/commit/454dfff6aa560dffad980a9258853108a44d5985)
+  which converts the `todolist` example.
+
+* Simple, toy applications may be more difficult to wrap - consumers will not
+  be able to defer decisions about `Send+Sync` support and will instead need to
+  implement simple locking as demonstrated in [this commit](
+  https://github.com/mozilla/uniffi-rs/commit/454dfff6aa560dffad980a9258853108a44d5985).
+
+* Existing applications that are yet to consider how to make their
+  implementations `Send+Sync` can not be wrapped until they have.
+
+* The examples which aren't currently marked with the `[Threadsafe]` attribute
+  will become more complex as they will all need to implement and explain how
+  they achieve being `Send+Sync`.
+
+* The perception that its more difficult to wrap interfaces will lead to less
+  adoption of the tool.
+
+## Pros and Cons of the Options
+
+### [Option 1]
+
+* Good, because we don't break anyone.
+* Bad, because we believe non-`Send+Sync` interfaces aren't useful in the
+  real-world, but we would pay the maintenance cost as though they were.
+* Bad, because locking remains hidden and leaves the door open to the same
+  gun we have already shot ourselves in the foot with.
+
+### [Option 2]
+
+* Good, because it makes the implementation of desired features easier.
+* Good, because it removes a foot-gun and makes locking both explicit and
+  visible to the developers of the uniffi-wrapped component.
+* Bad, because it breaks existing external consumers - it also breaks a couple
+  of internal consumers (for example, [fxa-client](
+  https://github.com/mozilla/application-services/blob/f3f0cf6e3386bf3036b074dad3950389cbd05746/components/fxa-client/src/fxa_client.udl#L97)),
+  but we believe fixing them is easy and low cost.
+
+## Implications
+
+We know there are external consumers and we know that this will break them.
+Therefore, we will commit to the following actions:
+
+* Communicating both this decision and how consumers can work around it as soon
+  as possible.
+
+* Noisily deprecate non-`Send+Sync` interfaces so existing consumers are likely
+  to see warnings and a link to our documentation when they upgrade.
+
+* Upgrade all internal mozilla consumers as soon as possible so they do not
+  issue deprecation warnings. As an example of what this entails,
+  see [this PR](https://github.com/mozilla/uniffi-rs/commit/454dfff6aa560dffad980a9258853108a44d5985)
+  which converts the `todolist` example to be `Send+Sync`.
```

Comment:
This migration plan might be useful to flesh out a bit. 

Consumers can get out ahead of these changes by making their interfaces `Threadsafe` before this change ships.
Do we need to warn consumers (e.g. having a transitionory release) before the switch?

Reply:
Do we know who all the consumers are?  I personally know of one non-Mozilla project using UniFFI (or at least did a month or two), and UniFFI was featured in "This Week in Rust" some time back...

Reply:
I've tried to clarify this a little more to indicate there will at *at least* 1 release where deprecation warnings will be issued but where non-threadsafe-marked interfaces are supported.

(When I wrote this I was trying to hedge my bets a little as I had a concern that supporting non-threadsafe interfaces would make doing some of this work very difficult, but that turns out to not be the case - we can probably keep non-threadsafe interfaces around for as long we feel reasonable for people to upgrade)